File: frontend/src/lib/r2.ts
import { S3Client, PutObjectCommand, GetObjectCommand } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';

function createR2Client() {
  const accountId = process.env.R2_ACCOUNT_ID;
  if (!accountId) throw new Error('R2_ACCOUNT_ID is not set');

  return new S3Client({
    region: 'auto',
    endpoint: `https://${accountId}.r2.cloudflarestorage.com`,
    credentials: {
      accessKeyId: process.env.R2_ACCESS_KEY_ID!,
      secretAccessKey: process.env.R2_SECRET_ACCESS_KEY!,
    },
  });
}

const BUCKET = () => {
  const b = process.env.R2_BUCKET_NAME;
  if (!b) throw new Error('R2_BUCKET_NAME is not set');
  return b;
};

export function assetKey(userId: string, generationId: string, filename: string): string {
  return `users/${userId}/generations/${generationId}/${filename}`;
}

export function thumbnailKey(key: string): string {
  const dot = key.lastIndexOf('.');
  return dot === -1 ? `${key}_thumb` : `${key.slice(0, dot)}_thumb${key.slice(dot)}`;
}

export async function uploadBuffer(
  key: string,
  buffer: Buffer,
  contentType: string,
): Promise<void> {
  const client = createR2Client();
  await client.send(
    new PutObjectCommand({
      Bucket: BUCKET(),
      Key: key,
      Body: buffer,
      ContentType: contentType,
    }),
  );
}

export async function downloadUrl(url: string): Promise<Buffer> {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Failed to download ${url}: ${res.status}`);
  return Buffer.from(await res.arrayBuffer());
}

export async function getSignedDownloadUrl(key: string, expiresInSeconds = 3600): Promise<string> {
  const client = createR2Client();
  const command = new GetObjectCommand({ Bucket: BUCKET(), Key: key });
  return getSignedUrl(client, command, { expiresIn: expiresInSeconds });
}

export function publicUrl(key: string): string {
  const base = process.env.R2_PUBLIC_URL;
  if (!base) throw new Error('R2_PUBLIC_URL is not set');
  return `${base.replace(/\/$/, '')}/${key}`;
}

File: backend/storage/r2.py
import boto3
import httpx
import os
from django.conf import settings

_r2 = None


def _client():
    global _r2
    if _r2 is None:
        _r2 = boto3.client(
            's3',
            endpoint_url=f'https://{settings.R2_ACCOUNT_ID}.r2.cloudflarestorage.com',
            aws_access_key_id=settings.R2_ACCESS_KEY_ID,
            aws_secret_access_key=settings.R2_SECRET_ACCESS_KEY,
            region_name='auto',
        )
    return _r2


def upload_bytes(key: str, data: bytes, content_type: str) -> None:
    _client().put_object(
        Bucket=settings.R2_BUCKET_NAME,
        Key=key,
        Body=data,
        ContentType=content_type,
    )


def download_url(url: str) -> bytes:
    with httpx.Client(timeout=120, follow_redirects=True) as client:
        r = client.get(url)
        r.raise_for_status()
        return r.content


def public_url(key: str) -> str:
    base = settings.R2_PUBLIC_URL.rstrip('/')
    return f'{base}/{key}' if base else ''


def asset_key(user_id, generation_id, filename: str) -> str:
    return f'users/{user_id}/generations/{generation_id}/{filename}'


def thumbnail_key(asset_key_str: str) -> str:
    base, _ = os.path.splitext(asset_key_str)
    return f'{base}_thumb.webp'
